x-pack/plugins/maps/public/components/toolbar_overlay/view.js

```diff
@@ -0,0 +1,282 @@
+import {
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiButtonIcon,
+  EuiPopover,
+  EuiContextMenu,
+  EuiSelectable,
+  EuiHighlight,
+  EuiTextColor,
+} from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import { getIndexPatternsFromIds } from '../../index_pattern_util';
+import _ from 'lodash';
+import { DRAW_TYPE } from '../../actions/store_actions';
+import { ES_GEO_FIELD_TYPE } from '../../../common/constants';
+
+const RESET_STATE = {
+  isPopoverOpen: false,
+  drawType: null
+};
+
+export class ToolbarOverlay extends React.Component {
+
+
+  state = {
+    isPopoverOpen: false,
+    uniqueIndexPatternsAndGeoFields: [],
+    drawType: null
+  };
+
+  _toggleToolbar = () => {
+    if (!this._isMounted) {
+      return;
+    }
+    this.setState(prevState => ({
+      isPopoverOpen: !prevState.isPopoverOpen,
+      drawType: null
+    }));
+  };
+
+  _closePopover = () => {
+    if (!this._isMounted) {
+      return;
+    }
+    this.setState(RESET_STATE);
+  };
+
+  _onIndexPatternSelection = (options) => {
+    if (!this._isMounted) {
+      return;
+    }
+
+    const selection = options.find((option) => option.checked);
+    const drawType = this.state.drawType;
+    this.setState(RESET_STATE, () => {
+      if (drawType) {
+        this.props.initiateDraw({ drawType: drawType, ...selection.value });
+      }
+    });
+  };
+
+  _activateDrawForFirstIndexPattern = (drawType) => {
+    if (!this._isMounted) {
+      return;
+    }
+    const indexPatternAndGeofield = this.state.uniqueIndexPatternsAndGeoFields[0];
+    this.setState(RESET_STATE, () => {
+      this.props.initiateDraw({ drawType: drawType, ...indexPatternAndGeofield });
+    });
+  };
+
+  componentDidMount() {
+    this._isMounted = true;
+  }
+
+  componentWillUnmount() {
+    this._isMounted = false;
+  }
+
+  async _getuniqueIndexPatternAndFieldCombos() {
+    const indexPatterns = await getIndexPatternsFromIds(this.props.uniqueIndexPatternIds);
+    try {
+      const uniqueIndexPatternsAndGeofields = [];
+      indexPatterns.forEach((indexPattern) => {
+        indexPattern.fields.forEach(field => {
+          if (field.type === ES_GEO_FIELD_TYPE.GEO_POINT || field.type === ES_GEO_FIELD_TYPE.GEO_SHAPE) {
+            uniqueIndexPatternsAndGeofields.push({
+              geoField: field.name,
+              geoFieldType: field.type,
+              indexPatternTitle: indexPattern.title,
+              indexPatternId: indexPattern.id
+            });
+          }
+        });
+      });
+      if (this._isMounted && !_.isEqual(this.state.uniqueIndexPatternsAndGeoFields, uniqueIndexPatternsAndGeofields)) {
+        this.setState({
+          uniqueIndexPatternsAndGeoFields: uniqueIndexPatternsAndGeofields
+        });
+      }
+    } catch(e) {
+      // swallow errors.
+      // the Layer-TOC will indicate which layers are disfunctional on a per-layer basis
+      return [];
+    }
+  }
+
+  componentDidUpdate() {
+    this._getuniqueIndexPatternAndFieldCombos();
```

Comment:
Why does `_getuniqueIndexPatternAndFieldCombos` need to get called on every state/prop change. How about just calling `_getuniqueIndexPatternAndFieldCombos` inside componentDidMount and inside _toggleToolbar when opening the popover?

Reply:
Visibility of the tool-button depends on whether there are filterable-layers to begin with, it also needs to be called at every update of `props.uniqueIndexPatternIds` (as layers are added). 

afaic, I would just leave it as-is, instead of trying to optimize this away.